### PR TITLE
Feature/tcomp 112 multiple runtimes and deps

### PR DESCRIPTION
**Pull request by sgandon**
(I modified this comment the 2/09/2016 to reflect changes made after Ryan's review)

Here are the modifications, that break certain APIs so I changed the version for most most of the modules.
- Created One interface *RuntimeInfo* to allow the Component developer to return the Runtime class name as well as the jar dependencies.
- removed all Runtime interface to move to on single *ComponentDefinition* method called
`RuntimeInfo getRuntimeInfo(Properties properties, ConnectorTopology connectorTopology);`. This method allow for the CompDev to compute the Runtime info according to the ComponentProperties and the Connector topology (see below).
- added a ConnectorTopoly enum so that the component designer can say what kind of connectivity it supports thanks to a new *ComponentDefinition* method called
`Set<ConnectorTopology> getSupportedConnectorTopologies();`. Here are the enum values (see the real file for complete javadoc).
```
public enum ConnectorTopology {
    OUTGOING,
    INCOMING,
    OUTGOING_AND_INCOMING,
    NONE;
}

```
- added a Runtime service in a new dedicated module with some submodules for an example/test case of a Component with 2 versions of the library
- I also changed the way the dependencies.txt file is generated to be used during test more easily so it implied a pom change for all components.

### CODE REVIEW

**Comment by sgandon** [comment]
Please do not merge until build is ok and the last warning (OSGI tests failure is resolved), But that would be nice you start looking at this and comment.

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/304/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/304/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/306/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/306/)

File: components-api-full-example/src/main/java/org/talend/components/fullexample/FullExampleDefinition.java [hunk 1]
```diff
@@ -62,7 +66,13 @@ public String getMavenArtifactId() {
...
 
     @Override
-    public Source getRuntime() {
+    public RuntimeInfo getRuntimeInfo(Properties properties, ConnectorTopology compType) {
         return null;
     }
+
+    @Override
+    public Set<ConnectorTopology> getSupportedConnectorTopologies() {
+        return EnumSet.of(ConnectorTopology.OUTPUTS);
```
Comment (RyanSkraba) [inline components-api-full-example/src/main/java/org/talend/components/fullexample/FullExampleDefinition.java]: Probably Collections.emptySet() ;)

File: components-api-runtime-service/components-api-runtime-service/pom.xml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    <execution>
+                        <id>osgi-bundle</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>bundle</goal>
+                        </goals>
+                        <configuration>
+                            <classifier>bundle</classifier>
+                            <instructions>
+                                <Bundle-SymbolicName>org.talend.multiruntime</Bundle-SymbolicName>
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/pom.xml]: This might be why the osgitest is failing -- there are a 5 bundles in this pull request with the symbolic name `org.talend.multiruntime!  Can we standardize this name to `${project.groupId}.${project.artifactId}`?

File: components-api-runtime-service/components-api-runtime-service/pom.xml [hunk 3]
```diff
@@ -0,0 +1,136 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/pom.xml]: Is this supposed to be an example module like the poc ones, or is this a pre-implementation of TCOMP-245?  

I'm not sure what this module is for, but it doesn't seem to be doing a lot!

File: components-osgitest/src/test/java/org/talend/components/api/PaxExamOptions.java [hunk 4]
```diff
@@ -76,15 +83,22 @@
                 mavenBundle().groupId("org.talend.components").artifactId("components-salesforce").classifier("tests")
                         .version(COMPONENTS_VERSION).noStart(),
                 mavenBundle("org.apache.servicemix.bundles", "org.apache.servicemix.bundles.hamcrest", "1.3_1"), //
+                mavenBundle().groupId("org.talend.components").artifactId("test-multiple-runtime-comp").classifier("bundle")
```
Comment (RyanSkraba) [inline components-osgitest/src/test/java/org/talend/components/api/PaxExamOptions.java]: My understanding of PaxExams is kind of limited -- do we need to add the multiple-runtime-comp and test-multiple-runtime-comp-v1/2 bundles for them to be loaded and resolved in the tests?

File: components-api/src/main/java/org/talend/components/api/component/runtime/RuntimeInfo.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * list all the depencencies required for this component to be executed at runtime
+     * 
+     * @param componentName name of the component to get the dependencies of.
+     * @return a set of maven uri following the pax-maven uri scheme @see
+     *         <a href="https://ops4j1.jira.com/wiki/display/paxurl/Mvn+Protocol">https://ops4j1.jira.com/wiki/display/paxurl/
+     *         Mvn+Protocol</a>
+     */
+    Set<URL> getMavenUrlDependencies();
```
Comment (RyanSkraba) [inline components-api/src/main/java/org/talend/components/api/component/runtime/RuntimeInfo.java]: I think order is significant here too (like in daikon).

File: components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import org.talend.daikon.sandbox.SandboxedInstance;
+
+public interface RuntimeService {
+
+    /**
+     * this will create a Runtime class based on the RuntimeInfo and using parentClassLoader if any is provided
+     * If you want to cast the returned object to some existing classes you are strongly advised to use the Properties classloader
+     * used to determine the runtimeInfo.
+     * This method also isolate the System properties for the current thread so you must call releaseRuntimeClass once the runtime
+     * is used in order to de-isolate the System properties from the current thread.
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java]: Think this comment needs to be updated :)  You mean {@link SandboxedInstance#close()}.

File: components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.talend.components.api.component.runtime.RuntimeInfo;
+import org.talend.daikon.sandbox.SandboxedInstance;
+
+public interface RuntimeService {
+
+    /**
+     * this will create a Runtime class based on the RuntimeInfo and using parentClassLoader if any is provided
+     * If you want to cast the returned object to some existing classes you are strongly advised to use the Properties classloader
+     * used to determine the runtimeInfo.
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java]: Confusing!  Did you mean that the parentClassLoader should be the one "closest" to the implementation of the component definition or properties that is creating the RuntimeInfo class?

File: components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                    } else {
+                        return null;
+                    }
+                }
+            });
+        }
+    }
+
+    @Override
+    public SandboxedInstance createRuntimeClass(RuntimeInfo runtimeInfo, ClassLoader classLoader) {
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeServiceImpl.java]: If the following two lines are the only value added by the RuntimeService module (along with the setup logic above, I guess), I suggest just sticking them beside the RuntimeInfo class in `components-api` or in `RuntimeUtil` in `component-api-service` if there are dependency issues.

File: components-api-runtime-service/components-api-runtime-service/src/main/resources/org/talend/components/multiruntime/messages.properties [hunk 9]
```diff
@@ -0,0 +1,5 @@
+component.MultiRuntimeComponent.title=File Reader
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/src/main/resources/org/talend/components/multiruntime/messages.properties]: looks like this file and the fileReader_icon32.png are left-overs and can be deleted.

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/pom.xml [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+					<execution>
+						<id>osgi-bundle</id>
+						<phase>package</phase>
+						<goals>
+							<goal>bundle</goal>
+						</goals>
+						<configuration>
+							<classifier>bundle</classifier>
+							<instructions>
+								<Bundle-SymbolicName>org.talend.multiruntime</Bundle-SymbolicName>
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/pom.xml]: Another org.talend.multiruntime!  We're rich with these!

Shouldn't this package be exporting and *not* importing org.talend.components.multiruntime?

Is there an issue with other bundles (notably test-multiple-runtimes-comp) exporting the same package.

This is so confusing to me...

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Override
+    public String getMavenGroupId() {
+        return "org.talend.components";
+    }
+
+    @Override
+    public String getMavenArtifactId() {
+        return "test-multiple-runtime-comp";
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java]: Wait, what?  Is this used at all now that we have multi-runtimes?

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+    public static RuntimeInfo getRuntimeInfoFromVersion(Version version) {
+        RuntimeInfo result = null;
+        switch (version) {
+        case VERSION_0_1:
+            result = new RuntimeInfo() {
+
+                @Override
+                public String getRuntimeClassName() {
+                    return "org.talend.multiruntime.MultiRuntimeComponentSource";
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java]: Just a quick note -- even in this poc, it would be a good idea to have different classnames depending on whether it's coming from v01 or v02.

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v01/src/main/resources/org/talend/multiruntime/messages.properties [hunk 13]
```diff
@@ -0,0 +1,4 @@
+component.MultiRuntimeComponent.title=File Reader
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v01/src/main/resources/org/talend/multiruntime/messages.properties]: These src/main/resources should *only* be in the `multiple-runtimes-comp`, right?

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v02/src/main/resources/org/talend/multiruntime/messages.properties [hunk 14]
```diff
@@ -0,0 +1,4 @@
+component.MultiRuntimeComponent.title=File Reader
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v02/src/main/resources/org/talend/multiruntime/messages.properties]: These src/main/resources should only be in the multiple-runtimes-comp, right?

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v02/pom.xml [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    <parent>
+        <groupId>org.talend.components</groupId>
+        <artifactId>components-parent</artifactId>
+        <version>0.15.0-SNAPSHOT</version>
+        <relativePath>../../../components-parent/pom.xml</relativePath>             
+    </parent>
+
+    <name>Components - MultiRuntimeComponent - runtime v02</name>
+    <groupId>org.talend.components</groupId>
+    <artifactId>test-multiple-runtime-comp-runtime-v02</artifactId>
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v02/pom.xml]: It's not a big deal, but I'd call this `multiple-runtime-comp-runtime-v02` (without the test).

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp/src/main/resources/org/talend/components/multiruntime/messages.properties [hunk 16]
```diff
@@ -0,0 +1,5 @@
+component.MultiRuntimeComponent.title=File Reader
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp/src/main/resources/org/talend/components/multiruntime/messages.properties]: These src/main/resources should only be in the multiple-runtimes-comp, right?

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp/src/test/java/org/talend/components/multiruntime/MultiRuntimeComponentTest.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.rules.ErrorCollector;
+import org.talend.components.api.service.ComponentService;
+import org.talend.components.api.service.internal.ComponentServiceImpl;
+import org.talend.components.api.test.SimpleComponentRegistry;
+import org.talend.components.multiruntime.MultiRuntimeComponentDefinition;
+
+@SuppressWarnings("nls")
+public class MultiRuntimeComponentTest {
```
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp/src/test/java/org/talend/components/multiruntime/MultiRuntimeComponentTest.java]: Unused?

File: components-api/src/main/java/org/talend/components/api/component/ComponentDefinition.java [hunk 18]
```diff
@@ -124,4 +128,25 @@
      */
     String getMavenArtifactId();
 
+    /**
+     * this will runtime the information required for running the component. They can depend an the given <code>properties
+     * </code> parameter and the component type required. The <code>componentType</code> shall be one of supported components
```
Comment (RyanSkraba) [inline components-api/src/main/java/org/talend/components/api/component/ComponentDefinition.java]: Some grammar, componentType -> connectorTopology.

File: components-osgitest/src/test/java/org/talend/components/api/MultiRuntimeTestITXX.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    public ComponentService getComponentService() {
+        BundleContext bundleContext = FrameworkUtil.getBundle(getClass()).getBundleContext();
+        try {
+            ServiceReference<?>[] compServiceRef = bundleContext.getAllServiceReferences(null, null);
+            for (ServiceReference<?> serviceRef : compServiceRef) {
+                System.out.println("SERVICE REF : " + serviceRef);
+                Object service = bundleContext.getService(serviceRef);
+                System.out.println("SERVICE : " + service);
```
Comment (RyanSkraba) [inline components-osgitest/src/test/java/org/talend/components/api/MultiRuntimeTestITXX.java]: confirmed that the symbolic name org.talend.multirun jar is not the one with the OSGI-INF that provides the MultiXxxComponentDefinition...

File: components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+package org.talend.components.api.component;
+
+import org.talend.components.api.component.runtime.RuntimeInfo;
+import org.talend.components.api.component.runtime.Sink;
+import org.talend.components.api.component.runtime.Source;
+
+/**
+ * This will define the possible types a component can be. I'll be used to retrieve the right {@link RuntimeInfo}.
+ */
+public enum ConnectorTopology {
```
Comment (RyanSkraba) [inline components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java]: Just a personal preference -- we're always stumbling over the terms INPUT and OUTPUT in the Studio.  It's hard to parse "The tSalesforce**Input** component supports **output** connections."

When we're describing connections relative to a component, can we use "INCOMING" and "OUTGOING" instead?

File: components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+//
+// ============================================================================
+package org.talend.components.api.component;
+
+import org.talend.components.api.component.runtime.RuntimeInfo;
+import org.talend.components.api.component.runtime.Sink;
+import org.talend.components.api.component.runtime.Source;
+
+/**
+ * This will define the possible types a component can be. I'll be used to retrieve the right {@link RuntimeInfo}.
```
Comment (RyanSkraba) [inline components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java]: And a component can be more than one type, depending on how it is connected in the job.   We never have the specific connections at our disposal when making this decision, just this enum.

File: components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * Component that only have output connectors. The runtime class must implement {@link Source}
+     */
+    OUTPUTS,
+    /**
+     * Component that only have input connectors. The runtime class must implement {@link Sink}
+     */
+    INPUTS,
+    /**
+     * Transform component that have both input and output connectors. The runtime class must implement ????
```
Comment (RyanSkraba) [inline components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java]: Just a note to think about -- transformation components will probably implement any of java.function.Function (if java 8), org.talend.daikon.java8.Function (if not),  Beam DoFn (when we have beam libraries available) or a PTransform (when we're committed to running in a Beam runner).

Also, farther in the future, any of these can return a Beam PTransform!

File: components-osgitest/src/test/java/org/talend/components/api/MultiRuntimeTestITXX.java (hunk 19 above)
Comment (RyanSkraba) [inline components-osgitest/src/test/java/org/talend/components/api/MultiRuntimeTestITXX.java]: Fixed!

File: components-api-full-example/src/main/java/org/talend/components/fullexample/FullExampleDefinition.java (hunk 1 above)
Comment (sgandon) [inline components-api-full-example/src/main/java/org/talend/components/fullexample/FullExampleDefinition.java]: yep

File: components-api-runtime-service/components-api-runtime-service/pom.xml (hunk 3 above)
Comment (sgandon) [inline components-api-runtime-service/components-api-runtime-service/pom.xml]: A discussed together, it is the beginning of something grand ;)
Kinding apart, this is the place where Vicnent work on REST apis for executing runtime input and output components will be placed too.

File: components-api/src/main/java/org/talend/components/api/component/runtime/RuntimeInfo.java (hunk 5 above)
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/runtime/RuntimeInfo.java]: yes you are right, I did not think so at first because of the default "depenendy.txt" implementation but you are right we may want to have something ordered here.

File: components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java (hunk 6 above)
Comment (sgandon) [inline components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java]: I will fix.

File: components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java (hunk 7 above)
Comment (sgandon) [inline components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeService.java]: I meant the properties classloader used to create the RuntimeInfo but I suppose the componentdefintion classloader will also do ;) .

File: components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeServiceImpl.java (hunk 8 above)
Comment (sgandon) [inline components-api-runtime-service/components-api-runtime-service/src/main/java/org/talend/components/runtimeservice/RuntimeServiceImpl.java]: I'll remove all this Service interface/impl/provider and create a simple RuntileUtil but keep is here as a placeholder for the future Runtime Rest API made by Vincent.

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java (hunk 11 above)
Comment (sgandon) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java]: very good point,I'll remove them.

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java (hunk 12 above)
Comment (sgandon) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java]: why ?

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v01/src/main/resources/org/talend/multiruntime/messages.properties (hunk 13 above)
Comment (sgandon) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v01/src/main/resources/org/talend/multiruntime/messages.properties]: yes indeed.

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v02/src/main/resources/org/talend/multiruntime/messages.properties (hunk 14 above)
Comment (sgandon) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/test-multiple-runtime-comp-runtime_v02/src/main/resources/org/talend/multiruntime/messages.properties]: yep

File: components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java (hunk 20 above)
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java]: I like it. let's do that then.

File: components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java (hunk 22 above)
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/ConnectorTopology.java]: I'll had the note in the javadoc.

File: components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java (hunk 12 above)
Comment (RyanSkraba) [inline components-api-runtime-service/components-poc-multiple-runtimes-and-deps/multiple-runtime-comp/src/main/java/org/talend/components/multiruntime/MultiRuntimeComponentDefinition.java]: It makes no difference to the runtime whether they have the same or different names, and it's easier when they are identifiable to the developer (when both are open in an IDE or searching, or when the classname is in an error message). 

As an alternative how about different package names?  `org.talend.multiruntime.v1.MultiRuntimeComponentSource` ?

File: components-api-runtime-service/components-api-runtime-service/pom.xml (hunk 3 above)
Comment (RyanSkraba) [inline components-api-runtime-service/components-api-runtime-service/pom.xml]: (We discussed in person, this LGTM as a placeholder for a runtime service to be fleshed out).

**Comment by RyanSkraba** [comment]
Hello!  LGTM.

Optional note remains at https://github.com/Talend/components/pull/224#discussion-diff-76984503R84 but I don't think it's important enough to block this PR!

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/380/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/380/)
